File: android/src/main/java/com/reactnativegooglesignin/RNOneTapSignInModule.java
package com.reactnativegooglesignin;

import static com.reactnativegooglesignin.RNGoogleSigninModule.rejectWithNullActivity;
import static com.reactnativegooglesignin.Utils.getUserProperties;

import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;
import android.util.Log;

import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.google.android.gms.auth.api.identity.BeginSignInRequest;
import com.google.android.gms.auth.api.identity.Identity;
import com.google.android.gms.auth.api.identity.SignInClient;
import com.google.android.gms.auth.api.identity.SignInCredential;
import com.google.android.gms.auth.api.signin.GoogleSignInStatusCodes;
import com.google.android.gms.common.api.ApiException;

@ReactModule(name = NativeOneTapSignInSpec.NAME)
public class RNOneTapSignInModule extends NativeOneTapSignInSpec {
    private static final int REQ_ONE_TAP = 2;

    private SignInClient oneTapClient;
    private PromiseWrapper promiseWrapper = new PromiseWrapper();

    public RNOneTapSignInModule(final ReactApplicationContext reactContext) {
        super(reactContext);
        reactContext.addActivityEventListener(new RNOneTapSigninActivityEventListener());
        oneTapClient = Identity.getSignInClient(reactContext);
    }

    public static final String NO_SAVED_CREDENTIAL_FOUND = "NO_SAVED_CREDENTIAL_FOUND";
    public static final String ONE_TAP_START_FAILED = "ONE_TAP_START_FAILED";

    @Override
    public void signIn(ReadableMap params, Promise promise) {
        Activity activity = getCurrentActivity();

        if (activity == null) {
            rejectWithNullActivity(promise);
            return;
        }
        promiseWrapper.setPromiseWithInProgressCheck(promise, "signIn");

        BeginSignInRequest request = Utils.buildOneTapSignInRequest(params);
        oneTapClient.beginSignIn(request)
            .addOnSuccessListener(activity, result -> {
                try {
                    activity.startIntentSenderForResult(result.getPendingIntent().getIntentSender(), REQ_ONE_TAP,
                        null, 0, 0, 0);
                } catch (IntentSender.SendIntentException e) {
                    Log.e(NativeOneTapSignInSpec.NAME, "Couldn't start One Tap UI: " + e.getLocalizedMessage());
                    promiseWrapper.reject(ONE_TAP_START_FAILED, e);
                }
            })
            .addOnFailureListener(activity, e -> {
                // No saved credentials found. Launch the One Tap sign-up flow, or
                // do nothing and continue presenting the signed-out UI.
                promiseWrapper.reject(NO_SAVED_CREDENTIAL_FOUND, e);
            });

    }

    @Override
    public void signOut(Promise promise) {
        oneTapClient
            .signOut().addOnSuccessListener(aVoid -> {
                promise.resolve(null);
            }).addOnFailureListener(e -> {
                promise.reject(e);
            });
    }

    private class RNOneTapSigninActivityEventListener extends BaseActivityEventListener {
        @Override
        public void onActivityResult(Activity activity, final int requestCode, final int resultCode, final Intent intent) {
             if (requestCode == REQ_ONE_TAP) {
                handleOneTapSignInResult(intent);
            }
        }
    }

    private void handleOneTapSignInResult(Intent data) {
        try {
            SignInCredential credential = oneTapClient.getSignInCredentialFromIntent(data);
            WritableMap userParams = getUserProperties(credential);
            promiseWrapper.resolve(userParams);
        } catch (ApiException e) {
            // TODO dupe
            int code = e.getStatusCode();
            String errorDescription = GoogleSignInStatusCodes.getStatusCodeString(code);
            promiseWrapper.reject(String.valueOf(code), errorDescription);
        }
    }
}

File: android/src/main/java/com/reactnativegooglesignin/RNOneTapSignInModule.kt
package com.reactnativegooglesignin

import android.app.Activity
import android.app.PendingIntent
import android.content.Intent
import android.content.IntentSender.SendIntentException
import androidx.core.app.ActivityCompat.startIntentSenderForResult
import androidx.credentials.ClearCredentialStateRequest
import androidx.credentials.CredentialManager
import androidx.credentials.CredentialOption
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import androidx.credentials.GetCredentialResponse
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.credentials.exceptions.GetCredentialException
import androidx.credentials.exceptions.NoCredentialException
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.BaseActivityEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.google.android.gms.auth.api.identity.AuthorizationResult
import com.google.android.gms.auth.api.identity.Identity
import com.google.android.gms.auth.api.signin.GoogleSignInStatusCodes
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.android.libraries.identity.googleid.GoogleIdTokenParsingException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class RNOneTapSignInModule(reactContext: ReactApplicationContext) :
  NativeOneTapSignInSpec(reactContext) {
  private val credentialManager: CredentialManager =
    CredentialManager.create(reactApplicationContext)
  private val addScopesPromiseWrapper = PromiseWrapper(NAME)
  private val oneTapUtils = OneTapUtils.OneTapUtils(reactApplicationContext)

  private val activityEventListener =
    object : BaseActivityEventListener() {
      override fun onActivityResult(
        activity: Activity?,
        requestCode: Int,
        resultCode: Int,
        data: Intent?
      ) {
        if (requestCode != REQUEST_AUTHORIZE) {
          return
        }
        when (resultCode) {
          Activity.RESULT_OK -> {
            try {
              val authorizationResult =
                Identity.getAuthorizationClient(activity!!).getAuthorizationResultFromIntent(data)
              val result = authResultToJsMap(authorizationResult)
              addScopesPromiseWrapper.resolve(result)
            } catch (e: Exception) {
              addScopesPromiseWrapper.reject(e)
            }
          }

          Activity.RESULT_CANCELED -> {
            addScopesPromiseWrapper.reject(
              GoogleSignInStatusCodes.SIGN_IN_CANCELLED.toString(),
              "User cancelled the operation."
            )
          }

          else -> {
            addScopesPromiseWrapper.reject("Failed to add scopes.")
          }
        }
      }
    }

  init {
    reactContext.addActivityEventListener(activityEventListener)
  }

  override fun invalidate() {
    reactApplicationContext.removeActivityEventListener(activityEventListener)
    super.invalidate()
  }

  override fun explicitSignIn(params: ReadableMap, promise: Promise) {
    val explicitRequest = oneTapUtils.buildExplicitOneTapSignInRequest(params)
    signInInternal(explicitRequest, promise)
  }

  override fun signIn(params: ReadableMap, promise: Promise) {
    val googleIdOption = oneTapUtils.buildOneTapSignInRequest(params)
    signInInternal(googleIdOption, promise)
  }

  private fun signInInternal(credentialOption: CredentialOption, promise: Promise) {
    val activity = currentActivity
    if (activity == null) {
      RNGoogleSigninModule.rejectWithNullActivity(promise)
      return
    }
    val request: GetCredentialRequest = GetCredentialRequest.Builder()
      .addCredentialOption(credentialOption)
      .build()

    CoroutineScope(Dispatchers.IO).launch {
      try {
        val result = credentialManager.getCredential(
          request = request,
          context = activity,
        )
        handleSignInSuccess(result, promise)
      } catch (e: GetCredentialException) {
        handleSignInError(e, promise)
      }
    }
  }

  private fun handleSignInError(e: GetCredentialException, promise: Promise) {
    when (e) {
      is GetCredentialCancellationException -> {
        promise.reject(GoogleSignInStatusCodes.SIGN_IN_CANCELLED.toString(), e.message, e)
      }

      is NoCredentialException -> {
        if (e.type == android.credentials.GetCredentialException.TYPE_NO_CREDENTIAL) {
          if (e.message?.contains("too many canceled sign-in prompts") == true) {
            promise.reject(ONE_TAP_START_FAILED, e.message, e)
          } else {
            promise.reject(NO_SAVED_CREDENTIAL_FOUND, e.message, e)
          }
        } else {
          promise.reject(e.type, e.message, e)
        }
      }

      else -> {
        promise.reject(e.type, e.message, e)
      }
    }
  }

  private fun handleSignInSuccess(result: GetCredentialResponse, promise: Promise) {
    val credential = result.credential

    when (credential) {
      is CustomCredential -> {
        if (credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
          try {
            val googleIdTokenCredential = GoogleIdTokenCredential
              .createFrom(credential.data)
            val userParams = oneTapUtils.getUserProperties(googleIdTokenCredential)
            promise.resolve(userParams)
          } catch (e: GoogleIdTokenParsingException) {
            promise.reject(NAME, e)
          }
        } else {
          promise.reject(NAME, "Unexpected type of custom credential")
        }
      }

      else -> {
        promise.reject(NAME, "Unexpected type of credential")
      }
    }
  }

  override fun requestAuthorization(params: ReadableMap, promise: Promise) {
    val activity = currentActivity
    if (activity == null) {
      RNGoogleSigninModule.rejectWithNullActivity(promise)
      return
    }
    // com.google.android.gms.common.Scopes can be used to get predefined scopes

    val authorizationRequest = oneTapUtils.buildAuthorizationRequest(params)

    Identity.getAuthorizationClient(activity)
      .authorize(authorizationRequest)
      .addOnSuccessListener { authorizationResult ->
        if (authorizationResult.hasResolution()) {
          // Access needs to be granted by the user
          val pendingIntent: PendingIntent = authorizationResult.pendingIntent!!
          try {
            addScopesPromiseWrapper.setPromiseWithInProgressCheck(promise, "addScopes")
            startIntentSenderForResult(
              activity,
              pendingIntent.intentSender,
              REQUEST_AUTHORIZE, null, 0, 0, 0, null
            )
          } catch (e: SendIntentException) {
            promise.reject("requestAuthorization", "Couldn't start Authorization UI: " + e.localizedMessage, e)
          }
        } else {
          val ret = authResultToJsMap(authorizationResult)
          promise.resolve(ret)
        }
      }
      .addOnFailureListener { e ->
        promise.reject("requestAuthorization", e)
      }
  }

  private fun authResultToJsMap(authorizationResult: AuthorizationResult): WritableMap {
    val accessToken = authorizationResult.accessToken
    val serverAuthCode = authorizationResult.serverAuthCode
    val grantedScopes = authorizationResult.grantedScopes

    val ret = Arguments.createMap()
    ret.putString("accessToken", accessToken)
    ret.putString("serverAuthCode", serverAuthCode)
    ret.putArray("grantedScopes", Arguments.fromList(grantedScopes))
    return ret
  }

  override fun signOut(promise: Promise) {
    CoroutineScope(Dispatchers.IO).launch {
      try {
        credentialManager.clearCredentialState(ClearCredentialStateRequest())
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("signOut", e)
      }
    }
  }

  companion object {
    const val ONE_TAP_START_FAILED = "ONE_TAP_START_FAILED"
    const val NO_SAVED_CREDENTIAL_FOUND = "NO_SAVED_CREDENTIAL_FOUND"
    const val REQUEST_AUTHORIZE = 9002
  }
}
